server/services/service.ts

```diff
@@ -0,0 +1,10 @@
+import type { Request } from 'express'
+
+export default class Service {
+  token: string
+
+  withTokenFromRequest(request: Request): typeof this {
```

Comment:
It's very easy to forget to call this method - and any place that forgets will default to using whatever token was stored by a user from the last request.  

One of the benefits of using typed languages is that you can use the compiler to make illegal states unrepresentable (or at least strongly encourage other developers to go down the right path!).

One way of doing this would be by using builder or factory, so services would no longer be singletons.
You could use the compiler to ensure that a valid service can only be instantiated with a token.

On the other hand, it would be much simpler if the token was considered as a required value and passed from the controller to the service as an additional parameter. This would reduce the complexity in your controllers and their tests - meaning you don't need to have special mocking code.

Reply:
For what it's worth - we did look at using continuation local storage (e.g [cls-hooked](https://github.com/Jeff-Lewis/cls-hooked)) and using as an equivalent to thread-local storage.

We'd extract tokens in some middleware and make them accessible directly in the rest clients, to avoid having to pass tokens about but it seemed a bit flaky and other libraries that also used CLS used to trounce over them.